### PR TITLE
Javelin Gun

### PR DESCRIPTION
**Pull request by Evankhell561**
It is a weapon that embeds projectiles in its victims so that the user can then exploit them to electrocute the victim
:cl:
add: Javelin Gun
/:cl:

### CODE REVIEW

**Review by NanakoAC — CHANGES_REQUESTED** [review]
This is a very rough draft which unfortunately leaves a lot to be desired.

The firing sound is way wrong, even for a placeholder it doesn't make sense, try to look for a bettter one. crossbow or pneumatic launcher perhaps

this doesn't make sense
https://i.imgur.com/CuwzGQz.png
A javelin is a spear, this is like saying knife sword. there's only one so no plurality is needed either

We have good sounds for things sticking into flesh, look at the unitology ritual blade and apply such sounds to the javelin embedding

The electrocution effect is supposed to be an AOE spray of lightning for several seconds, not just a couple of sparks. We have no custom sprites for this, and I was hoping you'd come up with something to improvise.  See what you can do with tesla sprite effects, otherwise we'll have to wait on more assets

Its supposed to continue for about 4 seonds, OR until the user lets go of the fire button. Being able to hold it down for extra crispiness is cathartically important

The detonation mode also frequently just doesn't work at all, not sure what's going on there

Also the detonation mode should consume a javelin (from the gun) to use. Counterintuitive I know, but its a game balance thing.

Lastly, the nailing effect is completely missing. On anything human sized, the javelin gun will launch them flying, and pin them to any walls they hit. Use push impulse for the launching, set a bump observation to watch for the victim hitting viable obstacles (walls mostly) and look at powered crossbow for how to pin someone to a wall

On a technical level it would probably also make sense not to implant the javelin into the victim's body, but have it as a seperate object on the turf. that'll make it easier to render and allow others to grab it to help pull it out

**Comment by Evankhell561** [comment]
image


image

**Comment by NanakoAC** [comment]
comments from functional testing:
The push effect is still missing. The javelin gun hits the target with enough force to launch them away, and pins them to any walls they hit in flight.  Use push impulse for this, check impulse.dm, and pin them if they're still flying when they hit a wall

The lightning effect sort of works, you got the controls right, but the functional effect of it is pretty minimal. I put a few slashers together and was unable to hurt either of them (epicentre or a nearby tile) with the electricity. Walking through it myself, i only took about 10 burn damage, it needs to be wayyyy stronger than this. For a guide on power, a javelin through the chest, and then a full discharge of electricity, should easily be enough to kill any T1 necromorph, and severely injure a T2. See if you can tweak the power to achieve these goals

Dropping an ammo casing on the floor doesn't exactly make sense. Having an object on the turf is a good idea if the victim gets pinned, in which case it should be mounted to the wall (see wallmount.dm), as well as rotated and offset to point into that wall. if the victim is too big to pin to anything, just embedding would be fine, in that case it doesn't make sense to leave an object behind outside of the target's body. 
since you've got two use cases here it may be best to make an extension to handle creating the electricity damage


Aside from the above, the audio is still not great. we don't have proper audio for it yet but there are definitely better placeholders, use the crossbow sound effects

File: code/game/atoms.dm [hunk 1]
```diff
@@ -633,4 +633,10 @@ its easier to just keep the beam vertical.
 	return FALSE
 
 /atom/proc/is_organic()
-	return FALSE
\ No newline at end of file
+	return FALSE
+
+/atom/proc/register_sharpnel(obj/item/weapon/material/shard/shrapnel/SP)
```
Comment (NanakoAC) [inline code/game/atoms.dm]: this is pronounced shrapnel

File: code/game/objects/items/weapons/material/shards.dm [hunk 2]
```diff
@@ -94,8 +94,76 @@
...
+		launcher = P.launcher
+		launcher.register_sharpnel(src)
 	..(loc, MATERIAL_STEEL)
 
 /obj/item/weapon/material/shard/phoron/New(loc)
 	..(loc, "phglass")
+
+/obj/item/weapon/material/shard/shrapnel/javeling
+	name = "javeling"
+	icon_state = "SpearFlight"
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/material/shards.dm]: its generally bad form to use capital letters in things like this

File: code/game/objects/items/weapons/material/shards.dm [hunk 3]
```diff
@@ -94,8 +94,76 @@
...
+	if(P)
+		launcher = P.launcher
+		launcher.register_sharpnel(src)
 	..(loc, MATERIAL_STEEL)
 
 /obj/item/weapon/material/shard/phoron/New(loc)
 	..(loc, "phglass")
+
+/obj/item/weapon/material/shard/shrapnel/javeling
+	name = "javeling"
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/material/shards.dm]: there's no g in javelin

File: code/game/objects/items/weapons/material/shards.dm [hunk 4]
```diff
@@ -94,8 +94,76 @@
...
+
+/obj/item/weapon/material/shard/shrapnel/javeling/examine(mob/user, distance)
+	. = ..()
+	if(launcher)
+		if(shock_count)
+			to_chat(user, SPAN_NOTICE("Its fully discharged."))
+		else
+			to_chat(user, SPAN_WARNING("Its charged with electricity."))
+
+/obj/item/weapon/material/shard/shrapnel/javeling/proc/remove_from_luncher_list()
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/material/shards.dm]: launcher not luncher

File: code/modules/projectiles/guns/projectile/javelin.dm [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	. = ..()
+	if(.)
+		return current_firemode.can_fire(target, user)
+
+/datum/firemode/automatic/shock
+
+/datum/firemode/automatic/shock/on_fire(atom/target, mob/living/user, clickparams, pointblank, reflex, obj/projectile)
+	if(istype(gun, /obj/item/weapon/gun/projectile/javelin_gun))
+		var/obj/item/weapon/gun/projectile/javelin_gun/J = gun
+		J.detone_javelin(user)
```
Comment (NanakoAC) [inline code/modules/projectiles/guns/projectile/javelin.dm]: detonate javelin

File: code/modules/projectiles/guns/projectile/javelin.dm [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	anchored = TRUE
+	pixel_x = -32
+	pixel_y = -32
+	var/life_time = 5
+	var/shock_damage = 35
+
+/obj/effect/overload/Initialize(mapload, n_life_time = 5)
+	. = ..()
+	life_time = n_life_time
+	START_PROCESSING(SSobj, src)
```
Comment (NanakoAC) [inline code/modules/projectiles/guns/projectile/javelin.dm]: This is too slow, use SSfastprocess instead, and break the damage down over 0.2 sec increments

File: code/modules/projectiles/projectile/bullets.dm [hunk 7]
```diff
@@ -241,5 +241,22 @@
...
 
-
-
+/obj/item/projectile/bullet/javelin
+	name = "javelin spears"
+	icon_state = "SpearFlight"
+	damage = 20
+	embed = TRUE
+	sharp = TRUE
+	armor_penetration = 50
```
Comment (NanakoAC) [inline code/modules/projectiles/projectile/bullets.dm]: make this 15

File: code/modules/projectiles/projectile/bullets.dm [hunk 8]
```diff
@@ -241,5 +241,22 @@
 	pixel_y = rand(-10,10)
 	..()
 
-
-
+/obj/item/projectile/bullet/javelin
+	name = "javelin spears"
+	icon_state = "SpearFlight"
+	damage = 20
```
Comment (NanakoAC) [inline code/modules/projectiles/projectile/bullets.dm]: 30 damage

**Review by NanakoAC — CHANGES_REQUESTED** [review]
first phase of review from code reading, some changes needed.

now going to do practical testing

File: code/modules/projectiles/projectile/bullets.dm (hunk 7 above)
Comment (Evankhell561) [reply code/modules/projectiles/projectile/bullets.dm]:
```suggestion
	armor_penetration = 15
```

File: code/modules/projectiles/projectile/bullets.dm (hunk 8 above)
Comment (Evankhell561) [reply code/modules/projectiles/projectile/bullets.dm]:
```suggestion
	damage = 30
```

**Comment by NanakoAC** [comment]
Forgot to unregister from the new system

`[13:01:29] Runtime in unsorted.dm,1071: Failed to stop processing. The overload (/obj/effect/overload) ([0x2032f3b]) (the deck) (210,139,4) (/turf/simulated/floor/tiled/dark) is being processed by SSfastprocess but de-queue attempt occured on SSobj.
  proc name: crash with (/proc/crash_with)
  src: null
  call stack:
  crash with("Failed to stop processing. The...")
  the overload (/obj/effect/overload): Destroy(0)
  qdel(the overload (/obj/effect/overload), 0)
  the overload (/obj/effect/overload): Process(2, 436, Fast Processing (/datum/controller/subsystem/processing/fastprocess))
  Fast Processing (/datum/controller/subsystem/processing/fastprocess): fire(0)
  Fast Processing (/datum/controller/subsystem/processing/fastprocess): ignite(0)
  Master (/datum/controller/master): RunQueue()
  Master (/datum/controller/master): Loop()
  Master (/datum/controller/master): StartProcessing(0)`

**Comment by Evankhell561** [comment]
> Forgot to unregister from the new system
> 
> `[13:01:29] Runtime in unsorted.dm,1071: Failed to stop processing. The overload (/obj/effect/overload) ([0x2032f3b]) (the deck) (210,139,4) (/turf/simulated/floor/tiled/dark) is being processed by SSfastprocess but de-queue attempt occured on SSobj. proc name: crash with (/proc/crash_with) src: null call stack: crash with("Failed to stop processing. The...") the overload (/obj/effect/overload): Destroy(0) qdel(the overload (/obj/effect/overload), 0) the overload (/obj/effect/overload): Process(2, 436, Fast Processing (/datum/controller/subsystem/processing/fastprocess)) Fast Processing (/datum/controller/subsystem/processing/fastprocess): fire(0) Fast Processing (/datum/controller/subsystem/processing/fastprocess): ignite(0) Master (/datum/controller/master): RunQueue() Master (/datum/controller/master): Loop() Master (/datum/controller/master): StartProcessing(0)`

sorry

**Comment by NanakoAC** [comment]
The accuracy is pretty bad, crank that up and set dispersion to 0 as well

**Comment by NanakoAC** [comment]
Freeing the victim from the pinning works, but its instant and throws a runtime error. 

It ought to have a do-after and take a few seconds, twice as long if the victim is trying to free themselves
`## ERROR: Cannot read null.extensions - extensions.dm - 108
## ERROR: proc name: remove extension (/proc/remove_extension)
  source file: extensions.dm,108
  usr: Bluespace Technician (/mob/living/carbon/human/bst)
  src: null
  usr.loc: the deck (211,138,4) (/turf/simulated/floor/tiled/dark)
  call stack:
remove extension(null, Mount (/datum/extension/mount/self_delete))
Mount (/datum/extension/mount/self_delete): dismount()
Mount (/datum/extension/mount/self_delete): mountpoint updated()
Mount (/datum/extension/mount/self_delete): on mountee move(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), the deck (211,139,4) (/turf/simulated/floor/tiled/dark), Bluespace Technician (/mob/living/carbon/human/bst))
Moved (/decl/observ/moved): raise event(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), the deck (211,139,4) (/turf/simulated/floor/tiled/dark), Bluespace Technician (/mob/living/carbon/human/bst))
Bluespace Technician (/mob/living/carbon/human/bst): Entered(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), the deck (211,139,4) (/turf/simulated/floor/tiled/dark))
Bluespace Technician (/mob/living/carbon/human/bst): Entered(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), the deck (211,139,4) (/turf/simulated/floor/tiled/dark))
Bluespace Technician (/mob/living/carbon/human/bst): Entered(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), the deck (211,139,4) (/turf/simulated/floor/tiled/dark))
the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling): forceMove(Bluespace Technician (/mob/living/carbon/human/bst))
the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling): forceMove(Bluespace Technician (/mob/living/carbon/human/bst))
...
Bluespace Technician (/mob/living/carbon/human/bst): equip to slot(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), 5, 1)
Bluespace Technician (/mob/living/carbon/human/bst): put in r hand(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling))
Bluespace Technician (/mob/living/carbon/human/bst): put in active hand(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling))
the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling): attack hand(Bluespace Technician (/mob/living/carbon/human/bst))
Bluespace Technician (/mob/living/carbon/human/bst): UnarmedAttack(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), 1)
Bluespace Technician (/mob/living/carbon/human/bst): LeftClickOn(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), "icon-x=75;icon-y=8;left=1")
Bluespace Technician (/mob/living/carbon/human/bst): ClickOn(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), "icon-x=75;icon-y=8;left=1")
the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling): Click("Turf", "infowindow.info", "icon-x=75;icon-y=8;left=1")
NanakoAC (/client): Click(the steel shrapnel (/obj/item/weapon/material/shard/shrapnel/javeling), "Turf", "infowindow.info", "icon-x=75;icon-y=8;left=1")
`

Also the push force on the impact is a bit lacking, can't really pin a slasher unless they're already next to the wall.
Crank it up majorly, a clean hit should launch a slasher across the room, and allow them to hit a wall about 5 tiles away

Feel free to adjust necromorph weights and push falloff code if necessary to make this work

**Comment by NanakoAC** [comment]
the audio placeholders are still bad, i'll try to get the right sounds from resident

**Comment by NanakoAC** [comment]
ah one more thing,
make the push force be halved if the victim is hit in a non-torso location. this is one weapon you want to aim centre mass with

**Comment by NanakoAC** [comment]
And throw this sound in whenever the javelin embeds/pins a victim, and when its pulled out:

`playsound(src, "fleshtear", VOLUME_MID, TRUE)`

**Comment by NanakoAC** [comment]
superceded by #1426

**Pull request closed without merging**